gnrc_sixlowpan_frag: initial import of Selective Fragment Recovery

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This provides an implementation for [6LoWPAN Selective Fragment Recovery](https://tools.ietf.org/html/draft-ietf-6lo-fragment-recovery-07) (SFR), an alternative fragmentation protocol for 6LoWPAN, which allows for end-to-end recovery and congestion control (currently not implemented) for fragmented IPv6 datagrams.
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
There is still a lot of more in-depth testing required including a dedicated test application (this is why is it still WIP), but you can compile `examples/gnrc_networking` with SFR already for a 6LoWPAN-capable board:

```
USEMODULE="gnrc_sixlowpan_frag_sfr" BOARD=samr21-xpro make -C examples/gnrc_networking flash term
```

and then ping another node also shipping with `gnrc_sixlowpan_frag_sfr`:

```
ping6 -s 1232 fe80::7b65:822:8693:9d5a
```

~~*[edit]: after posting this PR, I noticed, that there is a leak in the packet buffer after this... All the more reason why this is still WIP[/edit]*~~

There are also tests that should pass on a board of your choice:

```sh
BOARD=... make -C tests/gnrc_sixlowpan_frag_sfr -j flash term
```

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Requires ~~#12220~~ (merged), ~~#12559~~ (merged), ~~#12629~~, ~~#12845~~, and ~~#12848~~ (merged):

![Route to 6LoWPAN fragment forwarding](http://page.mi.fu-berlin.de/mlenders/sixlo_minfwd.svg)

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1179 @@
+                          _frag_seq(frag_desc), _frag_size(frag_desc),
+                          frag_desc->offset);
+                    gnrc_pktbuf_release(fbuf->pkt);
+                    fbuf->pkt = NULL;
```
Note to self: this leads to a duplicate release. fbuf is cleaned at error label.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1179 @@
+        else {
+            /* no ACK was requested and no error was causing an abort ACK*/
+            DEBUG("6lo sfr: no ACK requested by received fragment\n");
+            return;
```
I think the missing release before the return here is causing the leaks 😅

cgundogan commented on tests/unittests/tests-gnrc_sixlowpan_frag_vrb/tests-gnrc_sixlowpan_frag_vrb.c:
```diff
@@ -58,14 +57,13 @@ static uint8_t _out_dst[] = TEST_OUT_DST;
 static void set_up(void)
 {
     gnrc_sixlowpan_frag_vrb_reset();
-    tag = 0;
```
I am browsing through the test applications first. Was the removal of the `tag test` intended? If yes, why?

miri64 replied:
> Was the removal of the `tag test` intended? If yes, why?

This change to the test application is part of https://github.com/RIOT-OS/RIOT/pull/12559. With that PR the tag generator `gnrc_sixlowpan_frag_next_tag()` becomes independent of the `gnrc_sixlowpan_frag` module, so it does not need to be re-implemented, whenever the VRB is used without `gnrc_sixlowpan_frag` (as e.g. in this test or in SFR). The caveat is, that these tests don't know what the last tag was (they could assume it is just incremented, but it they shouldn't as this isn't necessarily future proof). Let's discuss this further in #12559 ;-).

> I am browsing through the test applications first. 

As stated in OP: this PR does not contain any test applications yet. All test applications or changes to test applications seen here are part of depending PRs.

miri64 commented:
Rebased to current to include https://github.com/RIOT-OS/RIOT/pull/12653 for `make debug` ;-)

miri64 commented:
Rebased to current master as #12220 and #12559 got merged.

miri64 commented:
Ok, I now can ping with huge datagrams in "rapid" (200ms as this is the estimate RTT) intervals without any crashes or leaks between two nodes. I'm not 100% sure resending is working as it should be. Further testing will show this hopefully. Next: actual forwarding.

miri64 commented:
Forwarding is now also working. There is still some weirdness happening with the ACKs and retries (I'm not sure if there is >1 retry made and the ACK timeout is restarted after it was fired even though the datagram is complete)... Will investigate this next week.

miri64 commented:
>  the ACK timeout is restarted after it was fired even though the datagram is complete

At least this I found now.

miri64 commented:
Rebased to current master

miri64 commented:
Rebased to current master and dependencies.

miri64 commented:
Added first round of tests.

miri64 commented:
I have full test coverage now (at least as far as I can judge without a coverage tester right now :sweat_smile:). On `native` they succeed, bur are still a bit flaky, on `iotlab-m3` and `samr21-xpro` they reliably fail. I will investigate what's up next. If that is found, I think I can remove the WIP label. @cgundogan may I squash before that?

miri64 commented:
> On `native` they succeed, bur are still a bit flaky, on `iotlab-m3` and `samr21-xpro` they reliably fail. I will investigate what's up next. If that is found, I think I can remove the WIP labe

And done. It was in the end as I suspected a timing issue with the ARQ timeout.

miri64 commented:
(no longer WIP, but still waiting on #12629)

miri64 commented:
Rebased and squashed to current master and dependencies.

miri64 commented:
(mainly to include #12759)

miri64 commented:
After many frustrations yesterday, I decided to clean up my code today a bit and found some issues with it. The results of the cleanup and repairs I now squashed and force pushed, as a series of fixup commits would have been too messy. The good news is: it works now, as far as I can judge... At least there are no crashes and I see a low number of successful pings. I suspect the low number of successes stemming from the issue demonstrated in #11256. The inter-frame gap helps a bit with mitigating it, but it isn't completely able to circumvent it. The ACKs only help to patch up the error, but in the end the datagram runs out of retries. I'm running some long-term pings tonight with the current configuration. Let's see what they will bring.

miri64 commented:
There is still an issue with the `interval` buffer of the (virtual) reassembly buffer running full (and are never properly released again) when the nodes are flooded. Partly I was able to fix it in #12848, however it isn't fixing all occurrences of that problem.

miri64 commented:
> There is still an issue with the `interval` buffer of the (virtual) reassembly buffer running full (and are never properly released again) when the nodes are flooded. Partly I was able to fix it in #12848, however it isn't fixing all occurrences of that problem.

Now, that I finally fixed that (see https://github.com/RIOT-OS/RIOT/pull/12848#issuecomment-561126573), I rebased and squashed once more, to have the latest master and dependencies. I will do another long-term test (hopefully https://unix.stackexchange.com/questions/490267/prevent-logoff-from-killing-tmux-session is properly fixed this time in my setup :sweat_smile:). In the meantime I will update for the most current draft version (there should be no functional changes, just some sections changed due to a review).

miri64 commented:
Long-term tests worked for the most part, however under high-load I found some leaks in the packet buffer. Will investigate.

miri64 commented:
> […] however under high-load I found some leaks in the packet buffer. Will investigate.

I think found the issue and the latest commit (12c2d92f39) should fix it. Additionally, I made it so, that abort pseudo-fragments request an ACK, unless after an ARQ timeout.

miri64 commented:
I run another test overnight to see if it is fixed.

miri64 commented:
That issue is fixed indeed, however after ~2.5h of heavy bombardment, pings suddenly stop to be successful (I get the occasional "error: packet buffer full" by the pinging node, which means it still tries to send [but runs out of buffer space, because un-acknowledged datagrams are hogging the packet buffer] and the pinged node even ACKs the pings fully, but it seems not be able to reply to the pings anymore (it just sends abort fragments which should not really happen). I'm currently investigating what is causing this, but given the timescale on which this happens this might take a while :-/.

miri64 commented:
Damn, I think it is the list of fragment descriptors that is degrading... On the pinged node I only have three entries left, though none of them are used.

miri64 commented:
I found one place were the fragment descriptor was not properly freed (after a fragment reached its retry limit, but the datagram still has some retries). I let it run once more to see if that did the trick!

miri64 commented:
What I can already say is, that the latest fixup increases performance significantly. Before, the lowest latency I was able to achieve with ping (so RTT) with payload size 1232 was >750ms. Now I see latencies (though rare) of ~500ms. Also it seems to me that far more pings receive a pong now.

miri64 commented:
(I can't really explain why... the latest fixup fixes a leak, so it should just not -- as it did before -- just stop working because of drained resources not become better...)

miri64 commented:
Worked! After another ~2.5h run instead of 3 remaining entries for the fragment description, I have all 64:

```
(gdb) print clist_count(_frag_descs_free)
$1 = 64
(gdb) print ARRAY_SIZE(_frag_descs_pool)
$2 = 64
```

\o/

miri64 commented:
Over the weekend, I will run a long term test pinging from two nodes (the original pinging node and the forwarder) and see what this will yield on Monday :-)

miri64 commented:
After some more long-term tests and with the latest fix-up I'm pretty confident that this can run for quiet a while without hang-ups or crashes now :-).

miri64 commented:
Squashed.

miri64 commented:
Rebased to current master as #12629 got merged.

miri64 commented:
Rebased to current master to resolve conflicts.

miri64 commented:
Made a change to make this PR not as hard to merge with #12879.

miri64 commented:
Going to rebase to current master. This PR has no dependencies anymore.

miri64 commented:
Done.

miri64 commented on sys/include/net/gnrc/sixlowpan/config.h:
```diff
@@ -58,8 +58,12 @@ extern "C" {
+#else   /* defined(MODULE_GNRC_SIXLOWPAN_FRAG_SFR) */
 #define CONFIG_GNRC_SIXLOWPAN_FRAG_FB_SIZE         (1U)
+#endif  /* defined(MODULE_GNRC_SIXLOWPAN_FRAG_SFR) */
 #endif
```
@leandrolanzieri @PeterKietzmann I guess I need to adapt this for Kconfig somehow? What do I need to do?

leandrolanzieri replied:
This behaviour in `sys/net/gnrc/network_layer/sixlowpan/frag/fb/Kconfig` could be:
```
config GNRC_SIXLOWPAN_FRAG_FB_SIZE
    int "Number of datagrams that can be fragmented simultaneously"
    default 4 if MODULE_GNRC_SIXLOWPAN_FRAG_SFR
    default 1
    help
        This determines the number of @ref gnrc_sixlowpan_frag_fb_t instances
        available.
```
Is that what you refer to?

miri64 replied:
Ok, but then I also need to provide the `gnrc_sixlowpan_frag_sfr` config first, right?

leandrolanzieri replied:
All the `MODULE_` symbols are generated from `USEMODULE`, so just by using `gnrc_sixlowpan_frag_sfr` module it will take 4 as default.

miri64 commented:
Added missing adaptations for current master.

miri64 commented:
Rebased and squashed due to conflicts caused by the merging of #12879. Also includes a fix for a bug I found during experimenting I forgot to push.

miri64 commented:
Rebased and squashed to current master.

miri64 commented:
Squashed and rebased to current master. There are some logical adaptations that this PR might need (`IS_USED()`, `IS_ACTIVE()`, Kconfig, ...). But other than that, this PR should be ready for review. The current implementation status was used for our experiments in [our last ACM ICN paper](https://dl.acm.org/doi/10.1145/3405656.3418719) with no issues.

miri64 commented:
The test still has some issues now after some rebases. I will investigate what went wrong. In the meantime: @leandrolanzieri could you have a look if I did the Kconfig integration correctly. Specifically, for the `app.config` in `tests/gnrc_sixlowpan_frag_sfr/` I wasn't sure if it was correct, as the values defined there do not show up in `make menuconfig`.

leandrolanzieri commented on tests/gnrc_sixlowpan_frag_sfr/app.config:
```diff
@@ -0,0 +1,8 @@
+CONFIG_KCONFIG_USEMODULE_GNRC_SIXLOWPAN_FRAG_SFR=y
+# disable router solicitations so they don't interfere with the tests
+CONFIG_GNRC_IPV6_NIB_NO_RTR_SOL=y
```
To set this you need to enable configuration of the module via Kconfig with
```
CONFIG_KCONFIG_USEMODULE_GNRC_IPV6_NIB=y
```

leandrolanzieri commented on tests/gnrc_sixlowpan_frag_sfr/app.config:
```diff
@@ -0,0 +1,8 @@
+# disable router solicitations so they don't interfere with the tests
+CONFIG_GNRC_IPV6_NIB_NO_RTR_SOL=y
+# preconfigure SFR for tests
+CONFIG_GNRC_SIXLOWPAN_SFR_OPT_WIN_SIZE=3
+CONFIG_GNRC_SIXLOWPAN_SFR_INTER_FRAME_GAP_US=5
+CONFIG_GNRC_SIXLOWPAN_SFR_MIN_ARQ_TIMEOUT_MS=100
+CONFIG_GNRC_SIXLOWPAN_SFR_OPT_ARQ_TIMEOUT_MS=100
```
For these, the configuration of GNRC 6LoWPAN should be enabled
```
CONFIG_KCONFIG_USEMODULE_GNRC_SIXLOWPAN=y
```

leandrolanzieri commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/Kconfig:
```diff
@@ -0,0 +1,91 @@
+        particular datagram will use.
+
+config GNRC_SIXLOWPAN_SFR_MIN_ARQ_TIMEOUT_MS
+    int "Default minimum amount of time in milliseconds a node should wait for an RFRAG Acknowledgment before it takes a next action (MinARQTimeOut)"
+    default 350
+
+config GNRC_SIXLOWPAN_SFR_MAX_ARQ_TIMEOUT_MS
+    int "Default maximum amount of time in milliseconds a node should wait for an RFRAG Acknowledgment before it takes a next action (MaxARQTimeOut)"
+    default 700
+
+config GNRC_SIXLOWPAN_SFR_OPT_ARQ_TIMEOUT_MS
+    int "Default starting point of the value of the amount of time in milliseconds that a sender should wait for an RFRAG Acknowledgment before it takes a next action (OptARQTimeOut)"
```
These prompts are really long and hard to read in the menuconfig interface. Would it be possible to split this into a prompt and help string?

miri64 replied:
I was thinking about this too. Even as a `@brief` they are quite un-brief ;-). They were mostly taken verbatim from the then current version of the draft, but that is not a must. Will see how I can fix them.

leandrolanzieri reviewed (commented):
>Specifically, for the app.config in tests/gnrc_sixlowpan_frag_sfr/ I wasn't sure if it was correct, as the values defined there do not show up in make menuconfig.

I commented the needed changes in the `app.config` file. I think that, with those lines, the configurations should be applied correctly.

miri64 commented:
Thanks for the comments @leandrolanzieri. Fixed.

miri64 commented:
> The test still has some issues now after some rebases.

Turns out, one of the [bugfixes applied for the paper](https://github.com/RIOT-OS/RIOT/commit/f228ab72cafb37282a6d3d42b3716d5db74dfad7) wasn't quite right and introduced another bug. Should be fixed now :-) with f70f7f8d33f45981afc4a28faf105d151221cf53 (using a new function introduced in 81f6032cf562e5801124070c97673e7bb5e5559b)

miri64 commented:
Ready for review. If preferred, I can squash this first.

benpicco commented:
Yes please squash first - there hasn't been a review yet.

miri64 commented:
Squashed

miri64 commented:
Arghs, forgot to reorder the commits I wanted to reorder ^^"

miri64 commented:
Done.

miri64 commented:
The draft is [now RFC 8931](https://www.rfc-editor.org/rfc/rfc8931.html), so I updated the doc accordingly

benpicco commented:
Needs a rebase - feel free to squash while doing so, no review has taken place yet.

miri64 commented:
> Needs a rebase […]

This will be fun... Not.

miri64 commented:
Rebased and tested both `tests/gnrc_sixlowpan_frag_minfwd` and `tests/gnrc_sixlowpan_frag_sfr` and if `examples/gnrc_networking` still compiles for `BOARD=samr21-xpro` without any modifications.

miri64 commented:
Fixed static errors.

miri64 commented:
I added `Makefile.ci` based on the once of `tests/gnrc_sixlowpan_frag_minfwd`. Since SFR is more complex than minimal forwarding it is to be expected, that those boards also won't fix (there will be more if any)

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+                }
+                reschedule_arq_timeout = true;
+            }
+            else if (_frag_ack_req(frag_desc)) {
```
Would be good to have a line to explain what those different cases are.

miri64 replied:
Done.

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+
+static inline kernel_pid_t _getpid(void)
+{
+    return IS_ACTIVE(TEST_SUITES) ? gnrc_sixlowpan_get_pid() : thread_getpid();
```
that could need some explanation

miri64 replied:
Added a comment.

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+    return IS_ACTIVE(TEST_SUITES) ? gnrc_sixlowpan_get_pid() : thread_getpid();
+}
+
+static void _clean_slate_datagram(gnrc_sixlowpan_frag_fb_t *fbuf)
```
So this resets the datagram to the state before sending?

miri64 replied:
Before fragmenting. SFR allows for resending the whole datagram if fragmentation timed out.

miri64 replied:
I added some doc to explain what this function does and when it can be used.

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+        gnrc_pktbuf_release(netif);
+        return NULL;
+    }
+    LL_PREPEND(res, netif);
```
Guess you could use the new gnrc_list functions here ;)

miri64 replied:
Done.

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+          gnrc_netif_addr_to_str(gnrc_netif_hdr_get_src_addr(netif_hdr),
+                                 netif_hdr->src_l2addr_len, addr_str),
+          hdr->base.tag);
+    if (gnrc_sixlowpan_frag_rb_exists(netif_hdr, hdr->base.tag)) {
+        DEBUG("6lo sfr: I am destination endpoint => adding to reassembly "
+              "buffer\n");
+        _try_reassembly(netif_hdr, pkt, offset, entry, page);
+    }
+    else if ((entry->entry.vrb = gnrc_sixlowpan_frag_vrb_get(
```
a short comment to explain the cases would be appreciated :)

miri64 replied:
Added

miri64 commented on sys/include/net/gnrc/netif/6lo.h:
```diff
@@ -18,12 +18,26 @@
+ * @brief   Selective Fragment Recovery
+ * @see     [RFC 8931](https://tools.ietf.org/html/rfc8931)
+ */
+#define GNRC_NETIF_6LO_LOCAL_FLAGS_SFR  (0x01)
```
Any interface capable of classic 6LoWPAN fragmentation should be capable of SFR. However, one might not activate this when e.g. other nodes in the network do not support SFR (it is not specified yet how this can be determined, but it most likely will use a flag 6LoWPAN capability option in the Router Advertisement).

benpicco replied:
so maybe call this Selective Fragment Recovery *enabled* then, then it's clearer it's a setting and now a hardware capability

miri64 replied:
None of these flags refer to capability, so I don't understand the confusion.

benpicco replied:
> Local 6LoWPAN capability flags

miri64 replied:
That comes from just reading Github and not your own doc ^^"

miri64 replied:
Mhhh now that I read my doc again: I named them capability flags, since that is what [the associated option](https://tools.ietf.org/html/rfc7400#section-3.3) is called. I try to clarify however, that these flags are less about hardware restraints and more about the network status quo.

benpicco replied:
I think changing the description to 

> Selective Fragment Recovery enabled

would already be enough to signal that this is a toggle

miri64 replied:
Sorry only saw your comment now. I added a longer description on the nature of these flags but also will add the verb as proposed.

miri64 commented on sys/Makefile.dep:
```diff
@@ -271,20 +271,26 @@ endif
   USEMODULE += gnrc_sixlowpan
-  USEMODULE += gnrc_sixlowpan_frag
+  ifeq (,$(filter gnrc_sixlowpan_frag_sfr,$(USEMODULE)))
+    USEMODULE += gnrc_sixlowpan_frag
```
That would be IMHO a separate PR. Either a dependent for this PR or as a follow-up, otherwise I believe the patch is too confusing. I would prefer a follow-up, as I don't know when I find time to work on that.

miri64 commented on sys/include/net/gnrc/sixlowpan/frag/sfr.h:
```diff
@@ -0,0 +1,161 @@
+/**
+ * @brief   Message type to signal an acknowledgement request timeout.
+ */
+#define GNRC_SIXLOWPAN_FRAG_SFR_ARQ_TIMEOUT_MSG     (0x0227)
+
+/**
+ * @brief   Message type to signal the sending of the next frame.
+ */
+#define GNRC_SIXLOWPAN_FRAG_SFR_INTER_FRAG_GAP_MSG  (0x0228)
```
Usually as a random number for the base and then consecutive numbers for the same module.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/Kconfig:
```diff
@@ -0,0 +1,100 @@
+
+config GNRC_SIXLOWPAN_SFR_MIN_FRAG_SIZE
+    int "Default minimum value for fragment size (MinFragmentSize)"
+    default 96
```
Yes, but congestion control might need to adopt this. Just a note: these config values all come from the RFC: https://www.rfc-editor.org/rfc/rfc8931.html#name-protocol-parameters

miri64 replied:
The names in parenthesis refer to the names in the RFC.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+    /**
+     * @brief   Acknowledgment request flag, sequence number, and fragment size
+     */
+    uint16_t ar_seq_fs;
```
Except for the ACK request flag, those fields are numeric values. How does a bitfield makes sense here?

benpicco replied:
```C
struct {
    uint16_t ar   : 1;
    uint16_t seq  : 5;
    uint16_t size : 10;
} uint16_t ar_seq_fs;
```

miri64 replied:
IIRC those kind of notifications are not portable which is why we don't use it in RIOT.

miri64 replied:
Ah, the problem is not portability (bitflags are part of the C11 standard and by now even MSP430 should be able to support C11), but size and byte order. I don't think the byte order will be like I want it to be (network byte order) when using the proposed solution, so I'd rather keep it as is.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+static char addr_str[GNRC_NETIF_HDR_L2ADDR_PRINT_LEN];
+#endif  /* MODULE_GNRC_IPV6_NIB */
+
+static evtimer_msg_t _arq_timer;
+static xtimer_t _if_gap_timer = { .offset = 0, .long_offset = 0 };
+static msg_t _if_gap_msg = { .type = GNRC_SIXLOWPAN_FRAG_SFR_INTER_FRAG_GAP_MSG };
+static uint32_t _last_frame_sent = 0U;
```
No, they are local to 6LoWPAN.

benpicco replied:
It's possible to have more then one 6lo interface (e.g. sub-GHz & 2.4 GHz) - wouldn't those compete about the timer, etc.?

miri64 replied:
The timer is used by 6LoWPAN alone and 6LoWPAN alone decides how to interpret those timers. I don't see how there can be a competition.

benpicco replied:
Ah yea you are right. I forgot that while we have multiple netdev threads, there is only one 6lo thread that will handle all interfaces.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+
+/* ====== INTERNAL FUNCTIONS USED BY PUBLIC FUNCTIONS ======
+ * ====== AND TO MANIPULATE INTERNAL DATA STRUCTURES  ====== */
+static inline uint32_t _bitmap_to_u32(uint8_t *bitmap)
```
Did not exist when I started implementing this, I believe. Will have a look.

miri64 replied:
Replaced the use of `_bitmap_to_u32()` with `unaligned_get_u32()` and removed  `_bitmap_to_u32()`.

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/Kconfig:
```diff
@@ -0,0 +1,100 @@
+
+config GNRC_SIXLOWPAN_SFR_INTER_FRAME_GAP_US
+    int "Default minimum amount of time between transmissions in microseconds (InterFrameGap)"
+    default 100
```
I wonder if this could be further optimized: I think we always get a TX done event, so we could measure the time it took to TX the frame and assume the next hop will take as much time for TXing it.

miri64 replied:
Sure, it is [mentioned in the RFC](https://www.rfc-editor.org/rfc/rfc8931.html#name-considerations-on-congestio), that the IFG is one of the factors that can be used for congestion control. Adapting this based on radio events, can be part of a congestion control strategy. It should however noted, that the IFG also plays into collision avoidance, so to take hidden terminal problems into account, it might not be enough to take the neighbor's TX time into account, but also its neighbors TX time too.

miri64 replied:
At any rate: this is more a thing for future adaptations, than for this PR.

miri64 commented on sys/include/net/gnrc/sixlowpan/frag/rb.h:
```diff
@@ -251,6 +251,25 @@ static inline void gnrc_sixlowpan_frag_rb_remove(gnrc_sixlowpan_frag_rb_t *rbuf)
+/* always true without TEST_SUITES defined */
+static inline bool gnrc_sixlowpan_frag_rb_ints_empty(void)
+{
+    return true;
```
As the restriction to the `TEST_SUITES` macro implies: this is a function mostly used for mocking.

miri64 replied:
(the function is only needed for testing, to check if the pool is empty, there is no requirement to check if the pool is empty when not testing)

miri64 replied:
I extended the comment above to clear up things.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/Kconfig:
```diff
@@ -0,0 +1,100 @@
+endif
+
+config GNRC_SIXLOWPAN_SFR_MIN_WIN_SIZE
+    int "Default minimum value of window size that the sender can use (MinWindowSize)"
+    default 1
```
When you want to tell the node not to stop reducing the window size lesser than that other value. The window size does not need to be bound to 1 at the lower end.

miri64 commented on sys/include/net/gnrc/sixlowpan/frag/sfr_types.h:
```diff
@@ -0,0 +1,64 @@
+                                 *   wait for an RFRAG Acknowledgment */
+    uint8_t cur_seq;            /**< Sequence number for next fragment */
+    uint8_t frags_sent;         /**< Number of fragments sent */
+    uint8_t window_size;        /**< Current window size */
```
Added the unit for the window size to the comment.

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/Kconfig:
```diff
@@ -0,0 +1,100 @@
+        @ref GNRC_SIXLOWPAN_SFR_MAX_FRAG_SIZE
+
+config GNRC_SIXLOWPAN_SFR_USE_ECN
+    bool "Indicates whether the sender should react to ECN (UseECN)"
```
Spelled it out in the help text as Explicit Congestion Notification (ECN).

miri64 commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1546 @@
+    return;
+error:
+    /* don't send abort for first fragment, the network does not know about
+     * the datagram */
```
Added a parameter for that to `_clean_up_fbuf()` and adapted its usage to set the proper status code when called.

miri64 commented on sys/include/net/gnrc/sixlowpan/frag/rb.h:
```diff
@@ -118,6 +132,28 @@ gnrc_sixlowpan_frag_rb_t *gnrc_sixlowpan_frag_rb_add(gnrc_netif_hdr_t *netif_hdr
+ *          exists in the first fragment.
+ *
+ * @return  true, if an entry with the given tuple exist.
+ * @return  false, if no entry with the given tuple exist.
```
Fixed.

miri64 commented:
Addressed comments

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1655 @@
+{
+    _generic_rb_entry_t entry = { .type = _UNDEF };
+
+    (void)page;
```
not needed

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1655 @@
+            evtimer_del((evtimer_t *)(&_arq_timer),
+                        &fbuf->sfr.arq_timeout_event.event);
+            fbuf->sfr.arq_timeout_event.msg.content.ptr = NULL;
+            if ((unaligned_get_u32(hdr->bitmap) == _null_bitmap.u32)) {
+                DEBUG("6lo sfr: fragmentation canceled\n");
+                _retry_datagram(fbuf);
+            }
```
When does that happen?

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1655 @@
+                                 netif_hdr->src_l2addr_len,
+                                 addr_str), hdr->base.tag,
+          hdr->bitmap[0], hdr->bitmap[1], hdr->bitmap[2], hdr->bitmap[3]);
+    if ((vrbe = gnrc_sixlowpan_frag_vrb_reverse(
+            gnrc_netif_hdr_get_netif(netif_hdr),
+            gnrc_netif_hdr_get_src_addr(netif_hdr),
+            netif_hdr->src_l2addr_len, hdr->base.tag)) != NULL) {
```
What case is this?

miri64 replied:
I also use DEBUG messages sometimes to comment on stuff... Should I really repeat what is stated there?

benpicco commented on sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c:
```diff
@@ -0,0 +1,1655 @@
+        gnrc_pktbuf_release(pkt);
+        return -ENOMEM;
+    }
+    else {
```
```suggestion
```
since you `return` in the case before, you can save yourself a level of indentation here.

benpicco reviewed (commented):
looks good, the added comments really help at better understanding what is going on here :) 
#13743 should benefit from this as it will fragment a lot.

miri64 commented:
Addressed the latest round of comments.

benpicco commented:
Somehow CI has an issue with `waspmote-pro` - might be related to #14799

miri64 commented:
Yeah, looks pretty unrelated. Should I rebase then. If so, may I squash in the same instance?

benpicco commented:
Yes, please squash (and rebase).

miri64 commented:
There was also a true issue, that I fixed with [the latest fixup](https://github.com/RIOT-OS/RIOT/pull/12648/commits/55313aeb1dd4a60fb51abca6681950f50dc7520e).

miri64 commented:
Squashed.

miri64 commented:
> Seems only to happen with netdev tests though, so might be caused by some dependency re-ordering.

IIRC there is no netdev-related dependency stuff happening in this PR. So this must be some weird side effect of unexpected netdev dependencies then ;-).

miri64 commented:
Mhhh the Murdock errors seem to be something related to `gnrc_netif` pulling in SFR headers, pulling in `evtimer`, pulling in `xtimer` and then `xtimer` having some messed up definitions. Not really sure exactly what is happening here... This used to work. Maybe @maribu or @kaspar030 can help.

miri64 commented:
Fixed static tests in the mean time.

miri64 commented:
The problem with `examples/nanocoap_server` was also legitimate. Fixed and squashed.

miri64 commented:
If that was @benpicco's intention: the problem is not fixed when merging https://github.com/RIOT-OS/RIOT/pull/15627 :-/

miri64 commented:
1. There was a rebase error, that un-guarded the `xtimer.h` include in `gnrc_netif.c`
2. If I `#if IS_USED(...)` guard the `sfr.h` include, `tests/driver_w5100` compiles for `waspmote-pro` now. It is only used to call the interface-specific init function of SFR, so I think the pre-processor magic is small enough. Still would be interesting why without 6LoWPAN and xtimer compiled in, ztimer-only boards have problems compiling such a combination...

maribu commented:
> 2. Still would be interesting why without 6LoWPAN and xtimer compiled in, ztimer-only boards have problems compiling such a combination...

The issue is that the ztimer-only boards cannot function with plain `xtimer`, due to incompatible clock frequency. To fix that, there is an 
```Makefile
ifneq (,$(filter xtimer,$(USEMODULE)))
  USEMODULE += xtimer_on_ztimer
endif
```

There is a compile time check via preprocessor that gets run when including `xtimer.h` that verifies that the `XTIMER_HZ` value is correctly set and compatible. Since boards incompatible with xtimer cannot provide a compliant value for them, this check will fail. However, if `xtimer_on_ztimer` is used, that check is skipped - `ztimer` can manage any timer frequency.

So the include to `xtimer.h` only works for those boards, if `xtimer_on_ztimer` is used - which is for those boards automatically done if `xtimer` is used.

miri64 commented:
> So the include to `xtimer.h` only works for those boards, if `xtimer_on_ztimer` is used - which is for those boards automatically done if `xtimer` is used.

So the conclusion is: if `xtimer.h` is included without the `xtimer` module compiled in, for those boards that are incompatible to `xtimer`, the compilation fails :-) (as `xtimer_on_ztimer` is not automatically included to fix the situation).

miri64 commented:
And some more bugs found by the compile tests fixed and squashed.

benpicco commented:
Did some testing (only single hop) and fragmentation works between two 6lo nodes with SFR enabled.
As expected fragmentation does not work between a SFR node and a 'classic' node, not sure if we should add that to the documentation. 

Two nodes with 'classic' fragmentation can still communicate as before.

miri64 commented:
> As expected fragmentation does not work between a SFR node and a 'classic' node, not sure if we should add that to the documentation.

Will add a note.

miri64 commented:
Done.

miri64 commented:
:tada: Thanks for the review and your input @benpicco
